### ISSUE TITLE
deseq-vst preprocessor recalculates train size factors manually in predict instead of using already saved train sizeFactors

### ISSUE THREAD

**Issue by hermidalc** [dncR/MLSeq#2]
You already calculate and save train sizeFactors here:
https://github.com/dncR/MLSeq/blob/9f2b484c0399a7d6057be896039ceee97b125535/R/classify.R#L244-L253

Though in predict you choose not to use them and redundantly recalculate train size factors manually:
https://github.com/dncR/MLSeq/blob/9f2b484c0399a7d6057be896039ceee97b125535/R/predict.R#L118-L124

**Comment by hermidalc** [comment]
Sorry I was reviewing the code paths again and I realize why you had to do this, since you cannot apply the per-sample sizeFactors from the training data as the sizeFactors for the test data as the samples are different.  Sorry about that!

**Comment by dncR** [comment]
Sorry for late reply. You are right. We calculate required parameters for test set samples while transforming and/or normalizing test samples by using train set parameters.

**Comment by hermidalc** [comment]
I saw how DESeq2 is doing it here https://github.com/mikelove/DESeq2/blob/master/R/core.R#L481-L514 and it made sense then what you've applied.  I understand now that you cannot use the sizeFactors from the training data because it's calculated along the dimension of training samples and therefore the vector isn't usable on test data where the number of samples and the samples themselves are different.  So you recaculate the geometric mean from training and apply it along the dimension of test samples.

**Comment by dncR** [comment]
Yeah. Actually, DESeq was not able to recalculate size factors of test set samples using train set parameters in the earlier versions. They have included it (and related functions) with recent versions. If we independently normalize test sets, overestimation problem will come up in the trained model but very low test set accuracy since it is normalized using different parameters than trained model.

**Comment by hermidalc** [comment]
> Yeah. Actually, DESeq was not able to recalculate size factors of test set samples using train set parameters in the earlier versions. They have included it (and related functions) with recent versions.

I don't see where in the latest DESeq2 reference manual what functionality they provide to do this?

**Comment by hermidalc** [comment]
One thing too about your implementation:
https://github.com/dncR/MLSeq/blob/9f2b484c0399a7d6057be896039ceee97b125535/R/predict.R#L118-L121
I did a simple test in R taking a training counts matrix and a test counts matrix (each with same number of genes) and applying those two lines of code to those results in NAs in the test size factors?  I believe it might be because you aren't filtering to make sure the log(counts.train) is finite and also filtering for counts.test > 0.  You can't assume that people have already filtered both counts matrices for low counts.

**Comment by dncR** [comment]
> > Yeah. Actually, DESeq was not able to recalculate size factors of test set samples using train set parameters in the earlier versions. They have included it (and related functions) with recent versions.
> 
> I don't see where in the latest DESeq2 reference manual what functionality they provide to do this?

This issue is not mentioned in the manual (at least I did not see such explanations), however, the description of argument `geoMeans` in function `DESeq2::estimateSizeFactors` is the answer. 

> geoMeans ---- by default this is not provided and the geometric means of the counts are cal- culated within the function. A vector of geometric means from another count matrix can be provided for a "frozen" size factor calculation

"frozen" means estimating size factors by using user-defined geometric means. We provide train set parameters via `geoMeans`.

**Comment by dncR** [comment]
> One thing too about your implementation:
> 
> https://github.com/dncR/MLSeq/blob/9f2b484c0399a7d6057be896039ceee97b125535/R/predict.R#L118-L121
> 
> 
> I did a simple test in R taking a training counts matrix and a test counts matrix (each with same number of genes) and applying those two lines of code to those results in NAs in the test size factors? I believe it might be because you aren't filtering to make sure the log(counts.train) is finite and also filtering for counts.test > 0. You can't assume that people have already filtered both counts matrices for low counts.

We add 1 to raw counts (by setting minimum count to 1) before downstream analysis. This strategy is frequently applied for classification tasks. In differential expression analysis, 0 counts are allowed and their effect have taken into account. I will check these lines if it causes crucial errors for downstream analyses when count matrix include zeros. In any case, this issue should be considered. Thanks for your contributions.
